### PR TITLE
Implementing Three-point Abel - another inverse Abel transformation algorithm

### PR DESCRIPTION
**Pull request by DhrubajyotiDas**
As discussed in #61, this pull request adds the 3-point Abel inversion technique to the repertoire of inversion algorithms available under PyAbel.

This technique was developed by [Dasch, 1992](https://www.osapublishing.org/ao/abstract.cfm?uri=ao-31-8-1146) (Applied Optics, Vol 31, No 8, March 1992, Pg 1146-1152). This algorithm exploits the observation that the value of the Abel inverted data at any radial position `r` is primarily determined from changes in the projection data in the neighborhood of `r`. 
From the paper: *the Abel integral is broken into segments around each point r<sub>j</sub>. In the neighborhood of each point,* the line-of-sight data *is expanded as a quadratic function of r - r<sub>j</sub>. Then each integral segment is integrated analytically. This method is called the 3-point Abel deconvolution since the quadratic expansion uses only the projections adjacent to r<sub>j</sub>.*

The original algorithm (with correction of typographical errors) was [implemented in Matlab](https://gist.github.com/DhrubajyotiDas/16d9381a78352f93bdfe) by Prof. Marshall Long's group at Yale. This PR is mostly a direct port of that code without any attempted optimizations (which are coming soon *hopefully*). 

In addition to the transform code, a new unit test checking the `three_point` transform against an analytical Gaussian is also included. The agreement between the two is essentially identical. 

    import numpy as np
    import matplotlib.pyplot as plt
    from abel.three_point import iabel_three_point

    n = 101
    r_max = 4
    r = np.linspace(-1*r_max, r_max, n)
    dr = r[1]-r[0]

    fig, axes = plt.subplots(ncols = 2)
    fig.set_size_inches(8,4)

    axes[0].set_xlabel("Lateral position, x")
    axes[0].set_ylabel("F(x)")
    axes[0].set_title("Original LOS signal")
    axes[1].set_xlabel("Radial position, r")
    axes[1].set_ylabel("f(r)")
    axes[1].set_title("Inverted radial signal")

    Mat = np.sqrt(np.pi)*np.exp(-1*r*r)
    AnalyticAbelMat = np.exp(-1*r*r)
    DaschAbelMat = iabel_three_point(Mat, 50, dr)

    axes[0].plot(r, Mat, 'r', label = r'$\sqrt{\pi}e^{-r^2}$')
    axes[0].legend()

    axes[1].plot(r, AnalyticAbelMat, 'k', lw = 1.5, alpha = 0.5, label = r'$e^{-r^2}$')
    axes[1].plot(r, DaschAbelMat, 'r--', lw = 1.5, label = '3-pt Abel')

    box = axes[1].get_position()
    axes[1].set_position([box.x0, box.y0, box.width * 0.8, box.height])
    axes[1].legend(loc='center left', bbox_to_anchor=(1, 0.5))
    fig.tight_layout()

dasch 3pt abel

### CODE REVIEW

**Comment by DanHickstein** [comment]
@DhrubajyotiDas, thanks for implementing this! We really have a nice comparison of different methods now. 

The agreement with the analytical solution is achieved without a scaling factor?

**Comment by DhrubajyotiDas** [comment]
Yes, the agreement is without any arbitrary scaling factor (apart from the necessary `dr` scaling)

**Comment by rth** [comment]
Looks great! I think we can merge this!

File: abel/three_point.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+        inv_IM[i] = F
+
+    return inv_IM
+
+def OP_D(i,j):
+    """
+    Calculate three-point abel inversion operator Di,j
+    The formula followed Dasch 1992 (Applied Optics) which contains several typos.
+    One correction is done in function OP1 follow Martin's PhD thesis
```
Comment (DanHickstein) [inline abel/three_point.py]: Can you provide a link to this thesis somewhere?

**Comment by DanHickstein** [comment]
I think that this looks pretty good as far as I can tell (though I can't say that I really understand exactly what's going on). 

While it's fresh in your mind, can you provide some comments in the code to explain what parts are doing which steps? If you want to merge this and add the comments in a separate PR, that's also fine with me.

It sounds like you have fixed some typos that occurred in the original paper. We should write down exactly what equations we are using. A good place for this might be the wiki? I have been thinking that we can include some long-form general explanations here, including the background of the transform method, the math, and any helpful advice that the user should know about the transform. I started a page for the Three Point transform here: https://github.com/PyAbel/PyAbel/wiki/Three-point-transform

File: abel/three_point.py (hunk 1 above)
Comment (DhrubajyotiDas) [inline abel/three_point.py]: https://www.lib.utexas.edu/etd/d/2002/martinkm07836/martinkm07836.pdf

Karl Matthew Martin, UT Austin, 2002.

He mentions the typo issue on Pg 114, and gives the corrected version of some of the equations on the following pages (presumably from Dasch's 1992 errata). I've pasted an example below. Left: Martin's thesis, Right: Dasch's paper.

image

**Comment by DhrubajyotiDas** [comment]
I am not too familiar with this algorithm either, but from what I can surmise, the real magic is happening in the`D` matrix, which is encoding the arithmetic that needs to be performed on the line-of-sight data to extract the inverted data. In general, such arithmetic cannot be encoded as a linear operator, but in the simplified case of approximating the raw projection data as a quadratic fit in the neighborhood of every data point, Dasch performed the algebra and solved the inverse Abel integral
image
analytically. The results of this algebra are Eqs 6 and 7 in the 1992 paper, and the actual math is discussed in Appendix A. 

For the purposes of the three_point technique, there were two typos in the original paper which were corrected by Dasch/Martin. I've added links to Martin's thesis in the documentation since it is the "correct" source of some of the equations used. I believe the paper also had other typos, but they were regarding other transform techniques outside the scope of this code. 

I think the wiki is a great idea. I'll add that to my to-do list!

**Comment by stggh** [comment]
My interpretation is that the derivative function <math>P<sup>'</sup>(r) = dF/dy</math>, common to all inverse transform methods is estimated using a 3-point approximation (to the derivative), Eq. (A2). This results in 3 integrals (trapezoid sum) for the inverse Abel transform.

The `hansenlaw` and `direct` methods perform a crude, direct, evaluation of the derivative (@rth has his own derivative function, see `abel.math`). There may be something to be learned here.

**Comment by DhrubajyotiDas** [comment]
@stggh Thanks for the clarification. I've added your interpretation to the wiki page on this technique.

**Comment by stggh** [comment]
One minor downer, speed optimizations may be required, the O2- VMI test case fails to finish (within the realm of operator patience) - could be operator error

**Comment by rth** [comment]
Regarding derivatives, it is true that the derivative functions for F could be common to most implementations that do not hard-code it (in particular `direct` and `hansenlaw`).   In fact, `direct`  implementation takes in a `gradient` function for that effect. By default it is set to `abel.math.derivative`  which is identical to `np.gradient` (i.e. a 2nd order centred differences) except that it also works for non uniform sampling. 

With the direct implementation I have also been using the [Savitzky-Golay](https://github.com/luli/hedp/blob/master/hedp/math/derivative.py#L62) differentiator for noisy data (that has been added to scipy of late). In  Dasch (Eq. A2) the derivative seems to be calculated as a average of centered 1st and 2d order differences, which might also have to do something with filtering noise (not sure).  BTW, if someone is interested,  here is some more reading about noise robust differentiators: 
http://www.holoborodko.com/pavel/numerical-methods/numerical-derivative/smooth-low-noise-differentiators/

If execution time is a problem, @DhrubajyotiDas would you mind just adding a line

     res_iabel['Three_point']['tr'].append(
                Timer(lambda: iabel_three_point(x)).timeit(number=NREPEAT)/NREPEAT)

 in [`abel.benchmark` at line 60](https://github.com/PyAbel/PyAbel/blob/master/abel/benchmark.py#L60) with the corresponding import of `iabel_three_point` at the top of the `__init__` function, so it is easier to benchmark the run time when optimizing this code?

**Comment by DanHickstein** [comment]
@stggh, I like your idea of including all of the methods in a single example. If the three-point method is too slow to be usable for your 1000x1000 example image, you could downsample the image to provide a reasonable run-time. Obviously it won't be a completely accurate comparison of the methods...

@rth, very good point about the derivative being a common feature of many of the transform methods. I like how you've allowed the direct methods to take any differentiation function - cool idea! Can the user simple call ``iabel.direct(IM, derivaive=scipy.signal.savgol_filter(deriv=1))`` to obtain the transform with the Savitzky-Golay differentiation? This should be mentioned in the docstring, since it seems like a feature that many users will like.

**Comment by DhrubajyotiDas** [comment]
@rth I've added the line you requested to `abel.benchmark`. I've also added the ability to cache the `D` matrix (called the "three_point operator matrix" for now...) so it can be reused if a batch of images of the same width are being processed. I will also be looking into vectorizing some of the loops to see if that makes things faster.

@stggh I can confirm that trying to generating a 1001x1001 operator matrix takes an unacceptable amount of time right now, so it's not operator error.

**Comment by DhrubajyotiDas** [comment]
@stggh With the latest commit, you ought to be able to transform a 1001x1001 image fairly quickly.

**Comment by stggh** [comment]
@DhrubajyotiDas    looks very good.  

examples/example_all_O2.py

Unfortunately, no free quadrant for the three_point transformed image (I will need to change example to smaller image segments)

all

direct inverse ...
                    0.4 sec
onion inverse ...
                   28.7 sec
hansen law inverse ...
                   1.0 sec
basex inverse ...
                   1.3 sec
three_point inverse ...
                   0.5 sec

**Comment by stggh** [comment]
@rth I also like  Savitzky-Golay and agree the method is worth investigating further

**Comment by DanHickstein** [comment]
@DhrubajyotiDas, great idea to add the caching. Is this PR ready to merge then?

**Comment by rth** [comment]
@DanHickstein Yes, I think we should merge this before Steve's PR as there could potentially be some conflicts in `three_point.py`. 
For Savitzky-Golay, it could be something like,

     abel.iabel_direct(IM, dr=dr, 
             derivative=lambda x: scipy.signal.savgol_filter(x,
                                       window_length=9, polyorder=3,  deriv=1)/dr)
I will check it up further (the interface changed slightly when it was added to Scipy with respect to the option I copied from somewhere) and add an example to the docstring.

**Comment by DhrubajyotiDas** [comment]
Yes, this PR is ready to be merged. Thanks for all your helpful comments!

**Comment by DanHickstein** [comment]
Great work!

**Pull request merged**